libs/constrained-generators/src/Constrained/SumList.hs

```diff
@@ -0,0 +1,309 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Constrained.SumList where
+
+import Data.List.NonEmpty (NonEmpty (..))
+import qualified Data.List.NonEmpty as NE
+import System.Random (Random (..))
+import Test.QuickCheck (Gen, choose, shuffle, vectorOf)
+
+-- =======================================================
+-- Helper functions for genSizedList
+
+data Solution t = Yes (NonEmpty [t]) | No [String]
+  deriving (Eq)
+
+instance Show t => Show (Solution t) where
+  show (No xs) = "No" ++ "\n" ++ unlines xs
+  show (Yes xs) = "Yes " ++ show xs
+
+isNo :: Solution t -> Bool
+isNo (No _) = True
+isNo (Yes _) = False
+
+concatSolution :: Show t => t -> String -> t -> Int -> [Solution t] -> Solution t
+concatSolution smallest pName total count [] =
+  No
+    [ "The sample in pickAll was empty"
+    , "smallest = " ++ show smallest
+    , "pred = " ++ pName
+    , "total = " ++ show total
+    , "count = " ++ show count
+    ]
+concatSolution _ _ _ _ (Yes xs : more) = foldr accum (Yes xs) more
+  where
+    accum (Yes zs) (Yes ys) = Yes (zs <> ys)
+    accum (No _) ys = ys
+    accum (Yes zs) (No _) = Yes zs -- This branch is unreachable
```

Comment:
```suggestion
concatSolution _ _ _ _ (Yes xs : more) = foldr accum xs more
  where
    accum (Yes zs) ys = Yes (zs <> ys)
    accum (No _) ys = ys
```

Reply:
I don't think it is type correct. 
```
accum (Yes zs) ys = zs <> ys
zs :: NonEmpty a     ys :: Solution a

Reply:
oh, right, you can get rid of the `fmap`
```
concatSolution _ _ _ _ (Yes xs : more) = Yes $ foldr accum xs more
  where
    accum (Yes zs) ys = zs <> ys
    accum (No _) ys = ys

```

Reply:
See further comments